### ISSUE TITLE
different behaviors in cumsum between torch.compile mode and eager mode

### ISSUE THREAD

**Issue by zoux1a** [pytorch/pytorch#117623]
### 🐛 Describe the bug

It's about function `torch.cumsum`, leading to different behaviours between torch.compile mode and eager mode. However, there is no mistake when `out` is set to default. I believe the error occurred due to `out` being set as a tensor of a smaller type than `input`, causing issues during the automatic conversion.
```python
import torch
import torch.nn as nn

class Model(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, out, input):
        out = torch.cumsum(out=out, dim=0,input=input)        
        return out

x = torch.randint(-32768, 32767, [2, 8, 8, 2, 1, 3, 4, 8], dtype=torch.int16)
y = torch.rand([2, 8, 8, 2, 1, 3, 4, 8], dtype=torch.float32)
z = x.clone()

model = Model().to(torch.device('cpu'))
eag = model(x, y)
opt = torch.compile(model.forward, mode='reduce-overhead')(z, y)

same_val = torch.allclose(eag.to('cpu'), 
                            opt.to('cpu'), 
                            rtol=1e-3, atol=1e-3, 
                            equal_nan=True)
if same_val == False : 
        raise ValueError('diff value')
```
error stacktrace
```
raise ValueError 'diff value'
```

### Versions

PyTorch version: 2.3.0.dev20240106+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.17 (main, Jul 5 2023, 20:41:20) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-86-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2070
GPU 1: NVIDIA GeForce RTX 2070
GPU 2: NVIDIA GeForce RTX 2070
GPU 3: NVIDIA GeForce RTX 2070

Nvidia driver version: 535.104.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
CPU family: 6
Model: 63
Thread(s) per core: 2
Core(s) per socket: 8
Socket(s): 2
Stepping: 2
CPU max MHz: 3200.0000
CPU min MHz: 1200.0000
BogoMIPS: 4794.64
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
Virtualization: VT-x
L1d cache: 512 KiB (16 instances)
L1i cache: 512 KiB (16 instances)
L2 cache: 4 MiB (16 instances)
L3 cache: 40 MiB (2 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s): 1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX disabled
Vulnerability L1tf: Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-triton==2.2.0+e28a256d71
[pip3] torch==2.3.0.dev20240106+cu118
[pip3] torchaudio==2.2.0.dev20240106+cu118
[pip3] torchvision==0.18.0.dev20240106+cu118
[conda] cudatoolkit 11.8.0 h6a678d5_0 defaults
[conda] numpy 1.24.3 pypi_0 pypi
[conda] pytorch-triton 2.2.0+e28a256d71 pypi_0 pypi
[conda] torch 2.3.0.dev20240106+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20240106+cu118 pypi_0 pypi
[conda] torchvision 0.18.0.dev20240106+cu118 pypi_0 pypi

**Comment by mikaylagawarecki** [comment]
Was able to repro on `main`. Issue seems to be with the downcast to `int16` for out. 

With `torch.manual_seed(5)`, 

```
# eager result: tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=torch.int16)
# compile result: tensor([0, 1, 1, 1, 2, 3, 4, 4, 4, 5], dtype=torch.int16)
```

Without `out` argument, output is 
```
tensor([0.6064, 1.1710, 1.2434, 1.9028, 2.6178, 3.1971, 4.1780, 4.8282, 4.8848,
        5.8049])
```

which gives the compile result when we do `.to(torch.int16)`. Marking triage reveiw for eager silent correctness

**Comment by malfet** [comment]
Added `oncall:pt2` label, not sure if it still needs to be discussed during todays triage review meeting

**Comment by xmfan** [comment]
mismatch seems to happen starting aot_eager.

``` python
import torch

def forward_fn(out, input):
    return torch.cumsum(out=out, dim=0, input=input)

input = torch.tensor([0.9201, 0.1166], dtype=torch.float32)
eag = torch.tensor([0, 0], dtype=torch.int32)
opt_eager = torch.tensor([0, 0], dtype=torch.int32)
opt_aot_eager = torch.tensor([0, 0], dtype=torch.int32)
opt_inductor = torch.tensor([0, 0], dtype=torch.int32)

forward_fn(out=eag, input=input)
torch.compile(forward_fn, backend="eager")(out=opt_eager, input=input)
torch.compile(forward_fn, backend="aot_eager")(out=opt_aot_eager, input=input)
torch.compile(forward_fn, backend="inductor")(out=opt_inductor, input=input)

print(f"eag={eag}")
print(f"opt_eager={opt_eager}")
print(f"opt_aot_eager={opt_aot_eager}")
print(f"opt_inductor={opt_inductor}")

# stdout:
# eag=tensor([0, 0], dtype=torch.int32)
# opt_eager=tensor([0, 0], dtype=torch.int32)
# opt_aot_eager=tensor([0, 1], dtype=torch.int32)
# opt_inductor=tensor([0, 1], dtype=torch.int32)
```

**Comment by zou3519** [comment]
tentatively marking hi-pri for silent incorrectness

**Comment by soulitzer** [comment]
cc @peterbell10

**Comment by peterbell10** [comment]
The graph returned by aot autograd does the cumsum in float precision and then casts to `int16` whereas the eager implementation casts to `int16` first and then does the `cumsum`.

**Comment by soulitzer** [comment]
From talking with @bdhirsh offline, one thing we could do here is to just update the functionalization rule for cumsum to do a cast on the inputs before doing the out-of-place op, though it would also be good to know if this is a broader issue.

**Comment by soulitzer** [comment]
I'm going to lower the priority for this issue, because in general, casting from float to integer type is disallowed (see https://github.com/pytorch/pytorch/blob/main/docs/source/tensor_attributes.rst), so it is a bit weird to begin with that cumsum is supporting this case. Separately from this issue there can exist cases where eager produces more accurate results (if we're casting to a higher-precision dtype, casting before produces more accurate results because the computation would be done in high-precision), but I would say that if the user cared a lot about the computation dtype they would've passed in the dtype= argument to cumsum.

**Comment by malfet** [comment]
@BoyuanFeng yeah, difference in type promotion logic between eager op and decomposed one will results in those errors

**Comment by aorenste** [comment]
This issue still reproduces on PyTorch 2.12.0a0+git7984635.

(Analysis done by Claude.)
